File: Data Structures & Algorithms/generate-parentheses/submission-5.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        stack   = []
        res     = []

        def recursive(opens, closes):
            if opens == closes == n:
                res.append("".join(stack))
                return
            if opens < n:
                stack.append("(")
                recursive(opens+1,closes)
                stack.pop()
            if closes < opens:
                stack.append(")")
                recursive(opens,closes+1)
                stack.pop()
        recursive(0,0)
        return res

        

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-5.cpp
class Solution {

public:
    int largestRectangleArea(vector<int>& heights) {

        // entry: idx of entry
        stack<int> stk;
        int N = heights.size();
        int res = 0;
        for (int i = 0; i <= N ; i ++) {
            while (!stk.empty() && ( i == N|| heights[i] <= heights[stk.top()])) {
                int h = heights[stk.top()];
                stk.pop();
                int w = stk.empty() ? i : i - stk.top() -1;
                res = max(res,h*w);
            }
            stk.push(i);
        }
        return res;
    }
};
